File: backend/app/schemas/uspd.py
import uuid
from datetime import datetime

from pydantic import Field

from app.schemas.common import ORMModel


class UspdEntryOut(ORMModel):
    id: uuid.UUID
    parent_id: uuid.UUID | None = None
    object: str
    model: str | None = None
    device_eui: str | None = None
    ip: str | None
    username: str | None
    password: str = ""
    comment: str | None
    section: str | None = None
    sim_number: str | None = None
    sim_ip: str | None = None
    sim_iccid: str | None = None
    sim_pin: str = ""
    sim_puk: str = ""
    sort_order: int = 0


class UspdSiteOut(ORMModel):
    id: uuid.UUID
    name: str
    notes: str | None
    system_id: uuid.UUID | None = None
    entries: list[UspdEntryOut]
    created_at: datetime
    updated_at: datetime


class UspdSiteWrite(ORMModel):
    name: str = Field(min_length=1, max_length=255)
    notes: str | None = Field(None, max_length=50_000)


class UspdSiteUpdate(ORMModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    notes: str | None = Field(None, max_length=50_000)


class UspdEntryWrite(ORMModel):
    object: str | None = None
    model: str | None = None
    device_eui: str | None = None
    ip: str | None = None
    username: str | None = None
    password: str | None = None
    comment: str | None = None
    section: str | None = None
    parent_id: uuid.UUID | None = None
    sim_number: str | None = None
    sim_ip: str | None = None
    sim_iccid: str | None = None
    sim_pin: str | None = None
    sim_puk: str | None = None
    sort_order: int | None = None


class UspdHwModelOut(ORMModel):
    id: uuid.UUID
    name: str
    created_at: datetime


class UspdHwModelWrite(ORMModel):
    name: str = Field(min_length=1, max_length=128)


class UspdObsidianFileResult(ORMModel):
    filename: str
    site_name: str | None = None
    created: bool = False
    skipped: bool = False
    entries: int = 0
    error: str | None = None


class UspdObsidianImportOut(ORMModel):
    created: int
    skipped: int
    failed: int
    files: list[UspdObsidianFileResult]


class UspdSimExcelRowResult(ORMModel):
    sheet_row: int
    phone: str | None = None
    ip: str | None = None
    site_name: str | None = None
    status: str
    error: str | None = None


class UspdSimExcelImportOut(ORMModel):
    created: int
    skipped: int
    unmatched: int
    empty: int
    rows: list[UspdSimExcelRowResult]

File: frontend/src/api/uspd.ts
import { apiFetch } from "./client";

export type UspdEntryOut = {
  id: string;
  parent_id: string | null;
  object: string;
  model: string | null;
  device_eui: string | null;
  ip: string | null;
  username: string | null;
  password: string;
  comment: string | null;
  section: string | null;
  sim_number: string | null;
  sim_ip: string | null;
  sim_iccid: string | null;
  sim_pin: string;
  sim_puk: string;
  sort_order: number;
};

export type UspdSiteOut = {
  id: string;
  name: string;
  notes: string | null;
  system_id: string | null;
  entries: UspdEntryOut[];
  created_at: string;
  updated_at: string;
};

export type UspdSiteWrite = {
  name: string;
  notes?: string | null;
};

export type UspdEntryWrite = {
  object?: string | null;
  model?: string | null;
  device_eui?: string | null;
  ip?: string | null;
  username?: string | null;
  password?: string | null;
  comment?: string | null;
  section?: string | null;
  parent_id?: string | null;
  sim_number?: string | null;
  sim_ip?: string | null;
  sim_iccid?: string | null;
  sim_pin?: string | null;
  sim_puk?: string | null;
  sort_order?: number;
};

export async function listUspdSites(q = ""): Promise<UspdSiteOut[]> {
  const sp = new URLSearchParams();
  if (q.trim()) sp.set("q", q.trim());
  const qs = sp.toString();
  return apiFetch<UspdSiteOut[]>(`/api/v1/uspd${qs ? `?${qs}` : ""}`);
}

export async function createUspdSite(body: UspdSiteWrite): Promise<UspdSiteOut> {
  return apiFetch<UspdSiteOut>("/api/v1/uspd", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateUspdSite(id: string, body: Partial<UspdSiteWrite>): Promise<UspdSiteOut> {
  return apiFetch<UspdSiteOut>(`/api/v1/uspd/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteUspdSite(id: string): Promise<void> {
  await apiFetch<void>(`/api/v1/uspd/${id}`, { method: "DELETE" });
}

export async function uploadUspdImage(file: File): Promise<{ url: string }> {
  const fd = new FormData();
  fd.append("file", file);
  return apiFetch<{ url: string }>("/api/v1/uspd/upload", {
    method: "POST",
    body: fd,
  });
}

export async function createUspdEntry(siteId: string, body: UspdEntryWrite): Promise<UspdEntryOut> {
  return apiFetch<UspdEntryOut>(`/api/v1/uspd/${siteId}/entries`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateUspdEntry(
  siteId: string,
  entryId: string,
  body: UspdEntryWrite,
): Promise<UspdEntryOut> {
  return apiFetch<UspdEntryOut>(`/api/v1/uspd/${siteId}/entries/${entryId}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteUspdEntry(siteId: string, entryId: string): Promise<void> {
  await apiFetch<void>(`/api/v1/uspd/${siteId}/entries/${entryId}`, { method: "DELETE" });
}

export type UspdHwModelOut = {
  id: string;
  name: string;
  created_at: string;
};

export async function listUspdHwModels(): Promise<UspdHwModelOut[]> {
  return apiFetch<UspdHwModelOut[]>("/api/v1/uspd/models");
}

export async function createUspdHwModel(name: string): Promise<UspdHwModelOut> {
  return apiFetch<UspdHwModelOut>("/api/v1/uspd/models", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
}

export type UspdObsidianFileResult = {
  filename: string;
  site_name: string | null;
  created: boolean;
  skipped: boolean;
  entries: number;
  error: string | null;
};

export type UspdObsidianImportOut = {
  created: number;
  skipped: number;
  failed: number;
  files: UspdObsidianFileResult[];
};

export async function importUspdObsidian(files: File[]): Promise<UspdObsidianImportOut> {
  const form = new FormData();
  for (const file of files) {
    form.append("files", file);
  }
  return apiFetch<UspdObsidianImportOut>("/api/v1/uspd/import-obsidian", {
    method: "POST",
    body: form,
  });
}

export type UspdSimExcelRowResult = {
  sheet_row: number;
  phone: string | null;
  ip: string | null;
  site_name: string | null;
  status: string;
  error: string | null;
};

export type UspdSimExcelImportOut = {
  created: number;
  skipped: number;
  unmatched: number;
  empty: number;
  rows: UspdSimExcelRowResult[];
};

export async function importUspdSimExcel(file: File): Promise<UspdSimExcelImportOut> {
  const form = new FormData();
  form.append("file", file);
  return apiFetch<UspdSimExcelImportOut>("/api/v1/uspd/import-sim-excel", {
    method: "POST",
    body: form,
  });
}
